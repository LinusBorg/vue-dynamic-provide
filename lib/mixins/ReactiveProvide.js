import Vue from 'vue'
import { pick } from '#lib/utils'
import { warn } from '#lib/utils/warn'

export default function createReactiveProvideMixin({
  name,
  props = false,
  attrs = false,
  listeners = false,
  include = false,
  inheritAs = false,
} = {}) {
  if (!name) {
    warn(
      `[vue-reactive-provide]: No name property found on options object when trying to create mixin.
      the 'name' property is required.`
    )
    return
  }
  let internalDataName = ''
  return {
    beforeCreate() {
      internalDataName = `$_reactiveProvide-${name}-Data`
      this[internalDataName] = createReactiveObject() // just an empty object so far
    },
    provide() {
      return {
        [name]: this[internalDataName],
      }
    },
    inject: inheritAs ? { [inheritAs]: name } : undefined,
    computed: {
      [name]: function() {
        let result
        try {
          result = {
            // first inheritAs from the parent if defined as such
            ...pickIf(inheritAs, this[inheritAs] || {}),
            // then add our own properties
            ...pickIf(props, this.$props),
            ...pickIf(attrs, this.$attrs),
            ...pickIf(include || [], this),
            ...pickIf(listeners, this.$listeners),
          }
        } catch (_) {
          /* istanbul ignore next */
          result = {}
        }
        return result
      },
    },
    watch: {
      [name]: {
        immediate: true,
        handler(val = {}) {
          const data = this[internalDataName]
          Object.keys(val).forEach(key => {
            if (data.hasOwnProperty(key)) {
              data[key] = val[key]
            } else {
              Vue.set(data, key, val[key])
            }
          })
        },
      },
    },
  }
}

function pickIf(predicate, obj) {
  if (predicate === true || typeof predicate === 'string') return obj

  if (Array.isArray(predicate)) {
    return pick(obj, predicate)
  }

  if (typeof predicate === 'object') {
    const entries = Object.entries(predicate)
    return entries.reduce((acc, [key, newKey]) => {
      acc[newKey] = obj[key]
      return acc
    }, {})
  }

  return {}
}

function createReactiveObject() {
  return new Vue({
    data: () => ({ obj: {} }),
  }).obj
}
