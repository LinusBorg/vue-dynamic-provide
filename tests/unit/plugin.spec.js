import { mount, createLocalVue } from '@vue/test-utils'
import { createGlobalMixin } from '../../src/mixins/GlobalMixin'

const testProd = !!process.env.VUE_APP_TEST_PROD
const DynamicProvide = testProd
  ? require('../../dist/vue-dynamic-provide.common').default
  : require('../../src').default

function mountWithPlugin(component, options) {
  const _Vue = createLocalVue()
  _Vue.use(DynamicProvide)

  return mount(component, {
    ...options,
    localVue: _Vue,
  })
}

describe('The Plugin', () => {
  it('shows Version', () => {
    const pluginVersion = DynamicProvide.version
    const { version } = require('../../package.json')

    expect(pluginVersion).toBe(version)
  })

  it('createGlobalMixin: creates mixin with functional beforeCreate hook', () => {
    const $options = {
      computed: {},
      provide: {},
      inject: {},
      dynamicProvide: {
        name: 'test',
        include: ['msg'],
      },
    }
    const $once = jest.fn().mockName('$once')
    const $on = jest.fn().mockName('$on')
    const VueStub = {
      $options,
      $once,
      $on,
    }

    const mixin = createGlobalMixin()

    expect(mixin).toMatchObject({
      beforeCreate: expect.any(Function),
    })

    mixin.beforeCreate.call(VueStub)
    expect($once).toHaveBeenCalledWith('hook:created', expect.any(Function))
    expect($on).not.toHaveBeenCalled()
    expect($options).toMatchObject({
      computed: {
        test: expect.any(Function),
      },
      dynamicProvide: expect.any(Object),
      provide: expect.any(Function),
    })

    expect($options).toMatchSnapshot()
  })

  it('installs global beforeCreate mixin', () => {
    const mixin = jest.fn()
    const Vue = {
      mixin,
    }

    DynamicProvide.install(Vue)

    expect(mixin).toHaveBeenCalledWith(
      expect.objectContaining({
        beforeCreate: expect.any(Function),
      })
    )
  })

  it('Basic: provides a working injection to a child component', () => {
    const wrapper = mountWithPlugin(
      require('./resources/plugin/Basic.vue').default
    )

    const child = wrapper.find({ name: 'child' })

    expect(child.vm.test).toMatchObject({
      msg: 'Testmessage',
    })
  })

  it('Props/Attrs: provides props & attrs', () => {
    const wrapper = mountWithPlugin(
      require('./resources/plugin/Props.vue').default,
      {
        propsData: {
          name: 'Test',
          msg: 'Hello',
        },
        attrs: {
          title: 'Title',
        },
      }
    )

    const child = wrapper.find({ name: 'child' })

    expect(child.vm.test).toMatchObject({
      name: 'Test',
      msg: 'Hello',
      title: 'Title',
    })
  })

  it('Listeners: provides callbacks', () => {
    const handler = () => {}
    const wrapper = mountWithPlugin(
      require('./resources/plugin/Listeners.vue').default,
      {
        listeners: {
          change: handler,
        },
      }
    )

    const child = wrapper.find({ name: 'child' })

    expect(child.vm.test).toMatchObject({
      change: expect.any(Function),
    })
  })

  it('OptionAsFunction: works when options are defined as a function', () => {
    const wrapper = mountWithPlugin(
      require('./resources/plugin/OptionAsFunction.vue').default,
      {
        propsData: {
          injectName: 'test',
        },
      }
    )

    const child = wrapper.find({ name: 'child' })

    expect(child.vm.test).toMatchObject({
      msg: 'Testmessage',
    })
  })
})
