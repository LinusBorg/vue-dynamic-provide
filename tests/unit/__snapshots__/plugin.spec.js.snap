// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Plugin createGlobalMixin: creates mixin with functional beforeCreate hook 1`] = `
Object {
  "computed": Object {
    "test": [Function],
  },
  "inject": Object {},
  "provide": [Function],
  "reactiveProvide": Object {
    "include": Array [
      "msg",
    ],
    "name": "test",
  },
}
`;
